File: java/des(GUI)x/src/desfunctions/DesMain.java
package desfunctions;

import javafx.util.Pair;
import stringmanip.StringManipulations;
import textconversions.TextConversions;

public class DesMain {

    public static String encryptDecryptInputTextPrep(String inputText, DesMainFunctions.CipherMode cipherMode, DesMainFunctions.TextMode textMode)  {
        DesMainFunctions.debugOutput.append(System.lineSeparator());

        switch (cipherMode)
        {
            case ENCRYPT:
                DesMainFunctions.debugOutput.append("PlainText: ").append(inputText + System.lineSeparator());
                DesMainFunctions.debugOutput.append("PlainText Hex and Binary Representations:" + System.lineSeparator());
                break;
            case DECRYPT:
                DesMainFunctions.debugOutput.append("CipherText: ").append(inputText + System.lineSeparator());
                DesMainFunctions.debugOutput.append("CipherText Hex and Binary Representations:" + System.lineSeparator());
                break;
        }

        String inputTextBin = "";

        switch (textMode)
        {
            case STRING:
                String inputTextHex = TextConversions.convertStringToHex(inputText);
                inputTextBin = TextConversions.convertHexToBin(inputTextHex);
                DesMainFunctions.debugOutput.append("Hex: ").append(String.join(" ", StringManipulations.chunkString(inputTextHex, 2)) + System.lineSeparator());
                DesMainFunctions.debugOutput.append("Binary: ").append(String.join(" ", StringManipulations.chunkString(inputTextBin, 8)) + System.lineSeparator());
                break;
            case HEX:
                inputText = inputText.replaceAll("\\s", "");

                inputTextBin = TextConversions.convertHexToBin(inputText);
                DesMainFunctions.debugOutput.append("Hex: ").append(String.join(" ", StringManipulations.chunkString(inputText, 2)) + System.lineSeparator());
                DesMainFunctions.debugOutput.append("Binary: ").append(String.join(" ", StringManipulations.chunkString(inputTextBin, 8)) + System.lineSeparator());
                break;
        }

        DesMainFunctions.debugOutput.append(System.lineSeparator());

        return inputTextBin;
    }

    public static String keyTextPrep(String initialKey, DesMainFunctions.TextMode textMode) {
        DesMainFunctions.debugOutput.append(System.lineSeparator());

        DesMainFunctions.debugOutput.append("Initial Key: ").append(initialKey + System.lineSeparator());
        DesMainFunctions.debugOutput.append("Initial Key Hex and Binary Representations:" + System.lineSeparator());

        String initialKeyBin = "";

        switch (textMode)
        {
            case STRING:
                String initialKeyHex = TextConversions.convertStringToHex(initialKey);
                initialKeyBin = TextConversions.convertHexToBin(initialKeyHex);

                DesMainFunctions.debugOutput.append("Hex: ").append(String.join(" ", StringManipulations.chunkString(initialKeyHex, 2)) + System.lineSeparator());
                DesMainFunctions.debugOutput.append("Binary: ").append(String.join(" ", StringManipulations.chunkString(initialKeyBin, 8)) + System.lineSeparator());
                break;
            case HEX:
                initialKey = initialKey.replaceAll("\\s", "");

                initialKeyBin = TextConversions.convertHexToBin(initialKey);
                DesMainFunctions.debugOutput.append("Hex: ").append(String.join(" ", StringManipulations.chunkString(initialKey, 2)) + System.lineSeparator());
                DesMainFunctions.debugOutput.append("Binary: ").append(String.join(" ", StringManipulations.chunkString(initialKeyBin, 8)) + System.lineSeparator());
                break;
        }

        DesMainFunctions.debugOutput.append(System.lineSeparator());

        return initialKeyBin;
    }

    public static Pair<String, String> encrypt(String plainTextBin, String initialKeyBin, boolean swapLastRound) {
        final String cipherTextHex = TextConversions.convertBinToHex(DesMainFunctions.desPrep(DesMainFunctions.CipherMode.ENCRYPT, plainTextBin, initialKeyBin, swapLastRound));
        final String cipherTextStr = TextConversions.convertHexToString(cipherTextHex);

        final String chunkedCipherTextHexStr = String.join(" ", StringManipulations.chunkString(cipherTextHex, 2));

        DesMainFunctions.debugOutput.append("Final CipherText (Hex): ").append(chunkedCipherTextHexStr + System.lineSeparator());
        DesMainFunctions.debugOutput.append("Final CipherText (String): ").append(cipherTextStr + System.lineSeparator());

        return new Pair<>(chunkedCipherTextHexStr, cipherTextStr);
    }

    public static Pair<String, String> decrypt(String cipherTextBin, String initialKeyBin, boolean swapLastRound) {
        final String plainTextHex = TextConversions.convertBinToHex(DesMainFunctions.desPrep(DesMainFunctions.CipherMode.DECRYPT, cipherTextBin, initialKeyBin, swapLastRound));
        final String plainTextStr = TextConversions.convertHexToString(plainTextHex);

        final String chunkedPlainTextHexStr = String.join(" ", StringManipulations.chunkString(plainTextHex, 2));

        DesMainFunctions.debugOutput.append("Final PlainText (Hex): ").append(chunkedPlainTextHexStr + System.lineSeparator());
        DesMainFunctions.debugOutput.append("Final PlainText (String): ").append(plainTextStr + System.lineSeparator());

        return new Pair<>(chunkedPlainTextHexStr, plainTextStr);
    }

    public static class WrongMenuChoiceException extends Exception {
        @Override
        public String toString() {
            return this.getClass().getCanonicalName() + ": Invalid Choice!";
        }
    }

}

File: kotlin/des(GUI)/src/main/kotlin/com/example/demo/DESFunctions/DesMain.kt
package DESFunctions

import DESFunctions.desMainFunctions.CipherMode
import DESFunctions.desMainFunctions.TextMode
import DESFunctions.desMainFunctions.debugOutput

object DesMain {

    fun encryptDecryptInputTextPrep(inputText: String, cipherMode: CipherMode, textMode: TextMode): String {
        var inputText = inputText
        debugOutput.append(System.lineSeparator())

        when (cipherMode) {
            CipherMode.ENCRYPT -> {
                debugOutput.append("PlainText: $inputText${System.lineSeparator()}")
                debugOutput.append("PlainText Hex and Binary Representations:${System.lineSeparator()}")
            }
            CipherMode.DECRYPT -> {
                debugOutput.append("CipherText: $inputText${System.lineSeparator()}")
                debugOutput.append("CipherText Hex and Binary Representations:${System.lineSeparator()}")
            }
        }

        var inputTextBin = ""

        when (textMode) {
            TextMode.STRING -> {
                val inputTextHex = TextConversions.convertStringToHex(inputText)
                inputTextBin = TextConversions.convertHexToBin(inputTextHex)
                debugOutput.append("Hex: ${inputTextHex.chunked(2).joinToString(" ")}${System.lineSeparator()}")
                debugOutput.append("Binary: ${inputTextBin.chunked(8).joinToString(" ")}${System.lineSeparator()}")
            }
            TextMode.HEX -> {
                inputText = inputText.replace("\\s".toRegex(), "")

                inputTextBin = TextConversions.convertHexToBin(inputText)
                debugOutput.append("Hex: ${inputText.chunked(2).joinToString(" ")}${System.lineSeparator()}")
                debugOutput.append("Binary: ${inputTextBin.chunked(8).joinToString(" ")}${System.lineSeparator()}")
            }
        }

        debugOutput.append(System.lineSeparator())

        return inputTextBin
    }

    fun keyTextPrep(initialKey: String, textMode: TextMode): String {
        var initialKey = initialKey
        debugOutput.append(System.lineSeparator())

        debugOutput.append("Initial Key: $initialKey${System.lineSeparator()}")
        debugOutput.append("Initial Key Hex and Binary Representations:${System.lineSeparator()}")

        var initialKeyBin = ""

        when (textMode) {
            TextMode.STRING -> {
                val initialKeyHex = TextConversions.convertStringToHex(initialKey)
                initialKeyBin = TextConversions.convertHexToBin(initialKeyHex)

                debugOutput.append("Hex: ${initialKey.chunked(2).joinToString(" ")}${System.lineSeparator()}")
                debugOutput.append("Binary: ${initialKeyBin.chunked(8).joinToString(" ")}${System.lineSeparator()}")
            }
            TextMode.HEX -> {
                initialKey = initialKey.replace("\\s".toRegex(), "")

                initialKeyBin = TextConversions.convertHexToBin(initialKey)
                debugOutput.append("Hex: ${initialKey.chunked(2).joinToString(" ")}${System.lineSeparator()}")
                debugOutput.append("Binary: ${initialKeyBin.chunked(8).joinToString(" ")}${System.lineSeparator()}")
            }
        }

        debugOutput.append(System.lineSeparator())

        return initialKeyBin
    }

    fun encrypt(plainTextBin: String, initialKeyBin: String, swapLastRound: Boolean): Pair<String, String> {
        val cipherTextHex = TextConversions.convertBinToHex(desMainFunctions.desPrep(CipherMode.ENCRYPT, plainTextBin, initialKeyBin, swapLastRound))
        val cipherTextStr = TextConversions.convertHexToString(cipherTextHex)

        val chunkedCipherTextHexStr = cipherTextHex.chunked(2).joinToString(" ")

        debugOutput.append("Final CipherText (Hex): $chunkedCipherTextHexStr${System.lineSeparator()}")
        debugOutput.append("Final CipherText (String): $cipherTextStr${System.lineSeparator()}")

        return Pair(chunkedCipherTextHexStr, cipherTextStr)
    }

    fun decrypt(cipherTextBin: String, initialKeyBin: String, swapLastRound: Boolean): Pair<String, String> {
        val plainTextHex = TextConversions.convertBinToHex(desMainFunctions.desPrep(CipherMode.DECRYPT, cipherTextBin, initialKeyBin, swapLastRound))
        val plainTextStr = TextConversions.convertHexToString(plainTextHex)

        val chunkedPlainTextHexStr = plainTextHex.chunked(2).joinToString(" ")

        debugOutput.append("Final PlainText (Hex): $chunkedPlainTextHexStr${System.lineSeparator()}")
        debugOutput.append("Final PlainText (String): $plainTextStr${System.lineSeparator()}")

        return Pair(chunkedPlainTextHexStr, plainTextStr)
    }

    class WrongMenuChoiceException : Exception() {
        override fun toString(): String {
            return this.javaClass.canonicalName + ": Invalid Choice!"
        }
    }

}
